src/dbc.rs

```diff
@@ -1,96 +1,309 @@
+    pub fn as_true_input_bearer(&self) -> Result<TrueInput> {
+        self.as_true_input(&self.owner_base().secret_key()?)
+    }
+
+    /// Generate hash of this DBC
+    pub fn hash(&self) -> [u8; 32] {
         let mut sha3 = Sha3::v256();
 
-        sha3.update(&self.content.hash().0);
-        sha3.update(&self.transaction.hash().0);
+        sha3.update(&self.content.to_bytes());
+        sha3.update(&self.transaction.hash());
 
         for (in_key, (mint_key, mint_sig)) in self.transaction_sigs.iter() {
-            sha3.update(&in_key.0.to_bytes());
+            sha3.update(&in_key.to_bytes());
             sha3.update(&mint_key.to_bytes());
             sha3.update(&mint_sig.to_bytes());
         }
 
+        for sp in self.spent_proofs.iter() {
+            sha3.update(&sp.to_bytes());
+        }
+
         let mut hash = [0u8; 32];
         sha3.finalize(&mut hash);
         hash
     }
 
-    // Check there exists a DbcTransaction with the output containing this Dbc
-    // Check there DOES NOT exist a DbcTransaction with this Dbc as parent (already minted)
-    pub fn confirm_valid<K: KeyManager>(&self, verifier: &K) -> Result<(), Error> {
-        for (input, (mint_key, mint_sig)) in self.transaction_sigs.iter() {
-            if !self.transaction.inputs.contains(input) {
-                return Err(Error::UnknownInput);
-            }
+    // Check there exists a Transaction with the output containing this Dbc
+    // note: common validation logic is shared by MintNode::reissue() and Dbc::confirm_valid()
+    //
+    // note: for spent_proofs to validate, the mint_verifier must have/know the spentbook section's public key.
+    pub fn confirm_valid<K: KeyManager>(
+        &self,
+        base_sk: &SecretKey,
+        mint_verifier: &K,
+    ) -> Result<(), Error> {
+        TransactionValidator::validate(
+            mint_verifier,
+            &self.transaction,
+            &self.transaction_sigs,
+            &self.spent_proofs,
+        )?;
 
-            verifier
-                .verify(&self.transaction.hash(), mint_key, mint_sig)
-                .map_err(|e| Error::Signing(e.to_string()))?;
-        }
-        if self.transaction.inputs.is_empty() {
-            Err(Error::TransactionMustHaveAnInput)
-        } else if self.transaction_sigs.len() < self.transaction.inputs.len() {
-            Err(Error::MissingSignatureForInput)
-        } else if self.transaction.inputs != self.content.parents {
-            Err(Error::DbcContentParentsDifferentFromTransactionInputs)
-        } else if !self.transaction.outputs.contains(&self.owner()) {
+        let owner = self.owner_once(base_sk)?.public_key_blst();
+
+        if !self
+            .transaction
+            .outputs
+            .iter()
+            .any(|o| *o.public_key() == owner)
+        {
             Err(Error::DbcContentNotPresentInTransactionOutput)
         } else {
             Ok(())
         }
```

Comment:
Should we also call `self.confirm_provided_amount_matches_commitment(..)`?

Reply:
maybe, but that would be "invalid" in a different sense than other confirm_valid() errors.

Presently we can distinguish between:
1. Dbc is valid and amount secrets are valid
2. Dbc is valid and amount secrets are invalid
3. Dbc is invalid and amount secrete are invalid
4. Dbc is invalid and amount secrets are valid

Of these, I think 1 and 2 are most important.  Not much need to distinguish between 3 and 4 usually I guess.

One way to retain same granularity would be to return an Enum from confirm_valid() instead of just ().

Or maybe we just keep the present API.

Anyway, I will open an issue for this.

Reply:
created #127